lib/recursive-methods.rb

```diff
@@ -1,49 +1,95 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError, "Please pick positive integer" if n < 0
+  return 1 if n == 0
+  n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  s[-1] + reverse(s[0...-1])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+
+  i = -1
+  j = s.length
+
+  until i == j
+    i += 1
+    j -= 1
+    temp = s[i]
+    s[i] = s[j]
+    s[j] = temp
+  end
+  return s if i == j
+  reverse_inplace(s)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n == 0
+  2 + bunny(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  return false if s.length.odd?
+  return true if s == ""
+
+  i = -1
+  j = s.length
+
+  s[i] == "(" && s[j] == ")" ? true : false
+
+  until i == j
+    i += 1
+    j -= 1
+    return false if s["i"] == s["j"]
+  end
+
+
+  return nested(s)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def search(array, value, c = 0)
+
+  return false if array.length == c
+  return true if value == array[c]
+
+  return search(array, value, c + 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n
+# Space complexity: O(n)
 def is_palindrome(s)
```

Comment:
⚠️ 
This is really a looped method rather than a recursive solution.

Reply:
...